swift-sdk/Internal/InAppManager.swift

```diff
@@ -515,9 +383,145 @@ extension InAppManager : InAppSynchronizerDelegate {
+                                         urlOpener: self.urlOpener)
+        }
+    }
+    
+    private func createAction(fromUrlOrAction urlOrAction: String) -> IterableAction? {
+        if let parsedUrl = URL(string: urlOrAction), let _ = parsedUrl.scheme {
+            return IterableAction.actionOpenUrl(fromUrlString: urlOrAction)
+        } else {
+            return IterableAction.action(fromDictionary: ["type" : urlOrAction])
+        }
+    }
+    
+    private func initializeMessagesMap() {
+        let messages = persister.getMessages()
+        for message in messages {
+            messagesMap[message.messageId] = message
+        }
+    }
+
+    // From client side
+    private func removePrivate(message: IterableInAppMessage) {
+        ITBInfo()
+        
+        updateMessage(message, didProcessTrigger: true, consumed: true)
+        self.internalApi?.inAppConsume(message.messageId)
+        self.callbackQueue.async {
+            self.notificationCenter.post(name: .iterableInboxChanged, object: self, userInfo: nil)
+        }
+    }
+
+    private static func isExpired(message: IterableInAppMessage, currentDate: Date) -> Bool {
+        guard let expiresAt = message.expiresAt else {
+            return false
+        }
+        
+        return currentDate >= expiresAt
+    }
+    
+    fileprivate static func isValid(message: IterableInAppMessage, currentDate: Date) -> Bool {
+        return message.consumed == false && isExpired(message: message, currentDate: currentDate) == false
+    }
+    
+    weak var internalApi: IterableAPIInternal?
+
+    private var synchronizer: InAppSynchronizerProtocol // this is mutable because we need to set internalApi
+    private let displayer: InAppDisplayerProtocol
+    private let inAppDelegate: IterableInAppDelegate
+    private let urlDelegate: IterableURLDelegate?
+    private let customActionDelegate: IterableCustomActionDelegate?
+    private let urlOpener: UrlOpenerProtocol
+    private let applicationStateProvider: ApplicationStateProviderProtocol
+    private let notificationCenter: NotificationCenterProtocol
+    
+    private let persister: InAppPersistenceProtocol
+    private var messagesMap = OrderedDictionary<String, IterableInAppMessage>() // This is mutable
+    private let dateProvider: DateProviderProtocol
+    private let retryInterval: TimeInterval // in seconds, if a message is already showing how long to wait?
+    private var lastDismissedTime: Date? = nil
+    private let updateQueue = DispatchQueue(label: "UpdateQueue")
+    private let scheduleQueue = DispatchQueue(label: "ScheduleQueue")
+    private let syncQueue = DispatchQueue(label: "SyncQueue")
+    private let callbackQueue = DispatchQueue(label: "CallbackQueue")
+    private var lastSyncTime: Date? = nil
+    private let moveToForegroundSyncInterval: Double = 1.0 * 60.0 // don't sync within sixty seconds
+}
+
+extension InAppManager : InAppNotifiable {
+    func onInAppSyncNeeded() {
+        ITBInfo()
```

Comment:
This will log twice since we already call ITBInfo in synchronize.

Reply:
We can call synchronize from other places. This log will help us trace whether we are calling synchronize from this method or some other method.